### ISSUE TITLE
Issue can not play Google DFP HLS or MPD AdTag.

### ISSUE THREAD

**Issue by ZohaibGhafoor** [google/ExoPlayer#4722]
i am using adTag generated from Google DFP and using HLS and MPD format buy i always get this error.

AdTag:
https://pubads.g.doubleclick.net/gampad/ads?iu=/39243592/413851632&description_url=[placeholder]&env=vp&impl=s&correlator=&tfcd=0&npa=0&gdfp_req=1&output=vast&sz=640x480&unviewed_position_start=1

```
08-24 16:22:01.112 5990-5990/com.google.android.exoplayer2.demo E/EventLogger: internalError [2.13, 0.00, window=0, loadError]
    com.google.android.exoplayer2.source.ads.AdsMediaSource$AdLoadException: java.io.IOException: Failed to load ad group 0
        at com.google.android.exoplayer2.ext.ima.ImaAdsLoader.handleAdGroupLoadError(ImaAdsLoader.java:1103)
        at com.google.android.exoplayer2.ext.ima.ImaAdsLoader.onAdError(ImaAdsLoader.java:597)
        at com.google.obf.ix.a(IMASDK:8)
        at com.google.obf.ir.a(IMASDK:68)
        at com.google.obf.ir.a(IMASDK:103)
        at com.google.obf.ij.a(IMASDK:73)
        at com.google.obf.jd.f(IMASDK:213)
        at com.google.obf.jd.a(IMASDK:44)
        at com.google.obf.je.b(IMASDK:39)
        at com.google.obf.je$1.shouldOverrideUrlLoading(IMASDK:4)
        at android.webkit.WebViewClient.shouldOverrideUrlLoading(WebViewClient.java:77)
        at com.android.webview.chromium.Ap.shouldOverrideUrlLoading(WebViewContentsClientAdapter.java:75)
        at org.chromium.android_webview.AwContentsClient.shouldIgnoreNavigation(AwContentsClient.java:19)
        at org.chromium.android_webview.AwContentsClientBridge.shouldOverrideUrlLoading(AwContentsClientBridge.java:160)
        at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method)
        at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:9)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6649)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826)
     Caused by: java.io.IOException: Failed to load ad group 0
        at com.google.android.exoplayer2.ext.ima.ImaAdsLoader.handleAdGroupLoadError(ImaAdsLoader.java:1103) 
        at com.google.android.exoplayer2.ext.ima.ImaAdsLoader.onAdError(ImaAdsLoader.java:597) 
        at com.google.obf.ix.a(IMASDK:8) 
        at com.google.obf.ir.a(IMASDK:68) 
        at com.google.obf.ir.a(IMASDK:103) 
        at com.google.obf.ij.a(IMASDK:73) 
        at com.google.obf.jd.f(IMASDK:213) 
        at com.google.obf.jd.a(IMASDK:44) 
        at com.google.obf.je.b(IMASDK:39) 
        at com.google.obf.je$1.shouldOverrideUrlLoading(IMASDK:4) 
        at android.webkit.WebViewClient.shouldOverrideUrlLoading(WebViewClient.java:77) 
        at com.android.webview.chromium.Ap.shouldOverrideUrlLoading(WebViewContentsClientAdapter.java:75) 
        at org.chromium.android_webview.AwContentsClient.shouldIgnoreNavigation(AwContentsClient.java:19) 
        at org.chromium.android_webview.AwContentsClientBridge.shouldOverrideUrlLoading(AwContentsClientBridge.java:160) 
        at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method) 
        at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:9) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6649) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826) 
     Caused by: AdError [errorType: PLAY, errorCode: AdErrorCode [name: VAST_LINEAR_ASSET_MISMATCH, number: 403], message: Linear assets were found in the VAST ad response, but none of them matched the video player's capabilities.]
        at com.google.obf.ir.a(IMASDK:102)
        at com.google.obf.ij.a(IMASDK:73) 
        at com.google.obf.jd.f(IMASDK:213) 
        at com.google.obf.jd.a(IMASDK:44) 
        at com.google.obf.je.b(IMASDK:39) 
        at com.google.obf.je$1.shouldOverrideUrlLoading(IMASDK:4) 
        at android.webkit.WebViewClient.shouldOverrideUrlLoading(WebViewClient.java:77) 
        at com.android.webview.chromium.Ap.shouldOverrideUrlLoading(WebViewContentsClientAdapter.java:75) 
        at org.chromium.android_webview.AwContentsClient.shouldIgnoreNavigation(AwContentsClient.java:19) 
        at org.chromium.android_webview.AwContentsClientBridge.shouldOverrideUrlLoading(AwContentsClientBridge.java:160) 
        at org.chromium.base.SystemMessageHandler.nativeDoRunLoopOnce(Native Method) 
        at org.chromium.base.SystemMessageHandler.handleMessage(SystemMessageHandler.java:9) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6649) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826) 
08-24 16:22:01.112 5990-5990/com.google.android.exoplayer2.demo D/EventLogger: timelineChanged [2.13, 0.00, window=0, periodCount=1, windowCount=1, reason=DYNAMIC
      period [?]
      window [?, false, false]
    ]
08-24 16:22:01.124 5990-5990/com.google.android.exoplayer2.demo D/EventLogger: mediaPeriodCreated [2.14, 0.00, window=0, period=0]
    loading [2.14, 0.00, window=0, period=0, true]
08-24 16:22:01.127 5990-5990/com.google.android.exoplayer2.demo W/IMASDK: Invalid internal message, ignoring. Please make sure the Google IMA SDK library is up to date. Message: gmsg://afma.google.com/videoDisplay?type=unload&sid=5a5b6f54-71fe-4602-9fb7-99f85c5a62c8&dt=1535109721120
08-24 16:22:02.835 5990-5990/com.google.android.exoplayer2.demo D/EventLogger: timelineChanged [3.85, 0.00, window=0, period=0, periodCount=1, windowCount=1, reason=DYNAMIC
      period [128.27]
      window [128.27, true, false]
    ]
08-24 16:22:02.852 5990-5990/com.google.android.exoplayer2.demo D/EventLogger: decoderEnabled [3.87, 0.00, window=0, period=0, video]
    decoderEnabled [3.87, 0.00, window=0, period=0, audio]
08-24 16:22:02.853 5990-6091/com.google.android.exoplayer2.demo I/OMXClient: IOmx service obtained
08-24 16:22:02.853 5990-5990/com.google.android.exoplayer2.demo D/EventLogger: tracksChanged [3.87, 0.00, window=0, period=0, 
      Renderer:0 [
        Group:0, adaptive_supported=N/A [
          [X] Track:0, id=1, mimeType=video/avc, res=1280x720, supported=YES
        ]
      ]
      Renderer:2 [
08-24 16:22:02.854 5990-5990/com.google.android.exoplayer2.demo D/EventLogger:     Group:0, adaptive_supported=N/A [
          [X] Track:0, id=2, mimeType=audio/mp4a-latm, channels=2, sample_rate=44100, language=und, supported=YES
        ]
      ]
    ]
```

**Comment by andrewlewis** [comment]
The cause is `AdError [errorType: PLAY, errorCode: AdErrorCode [name: VAST_LINEAR_ASSET_MISMATCH, number: 403], message: Linear assets were found in the VAST ad response, but none of them matched the video player's capabilities.]`.

Searching for `VAST_LINEAR_ASSET_MISMATCH` may help. Or if you provide the ad tag we can take a look.

**Comment by ZohaibGhafoor** [comment]
@andrewlewis 
@andrewlewis 
AdTag:
https://pubads.g.doubleclick.net/gampad/ads?iu=/39243592/413851632&description_url=[placeholder]&env=vp&impl=s&correlator=&tfcd=0&npa=0&gdfp_req=1&output=vast&sz=640x480&unviewed_position_start=1

**Comment by ZohaibGhafoor** [comment]
@andrewlewis 
AdTag:
https://pubads.g.doubleclick.net/gampad/ads?iu=/39243592/413851632&description_url=[placeholder]&env=vp&impl=s&correlator=&tfcd=0&npa=0&gdfp_req=1&output=vast&sz=640x480&unviewed_position_start=1

**Comment by andrewlewis** [comment]
We're setting the supported MIME types on the `AdsRenderingSettings` but it looks like they aren't being propagated to the code that determines what ad media can be played. I've filed [Internal: b/113165344] to investigate and will update this when I hear back. Thanks for reporting this!

**Comment by andrewlewis** [comment]
A quick update on this: the IMA SDK team has implemented a fix, so hopefully this will go out in an IMA release soon and then we can update the IMA extension's dependency on the SDK to pick it up.

**Comment by ZohaibGhafoor** [comment]
@andrewlewis thankyou so much to give us an update news. now we are waiting for fix update.

**Comment by andrewlewis** [comment]
@ZohaibGhafoor Do you have a working ad tag for reproducing this issue? I am trying to verify that it's fixed with the latest IMA SDK 3.10.2 but the ad tag provided above no longer works.

**Comment by ZohaibGhafoor** [comment]
@andrewlewis we refreshed the adtag url which provided above please check.

**Comment by andrewlewis** [comment]
@ZohaibGhafoor This does seem to be fixed with IMA SDK 3.10.2. I'll plan to update this issue when we push the change to upgrade the dependency to the development branch, and it should make it into the next release.

If you need the fix urgently and are depending on ExoPlayer via sources you can update the relevant dependencies in `extensions/ima/build.gradle` to point to `com.google.ads.interactivemedia.v3:interactivemedia:3.10.2` and `com.google.android.gms:play-services-ads:17.1.1`, and add `<meta-data android:name="com.google.android.gms.ads.AD_MANAGER_APP" android:value="true"/>` in a new `<application>` element in `extensions/ima/src/main/AndroidManifest.xml`.

**Comment by andrewlewis** [comment]
a3c046983765b26688e8ac0e24a1539172594f7a updates the IMA SDK to 3.10.2.
